File: src/Core/Validator.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

// For easy access to brushes
using System.Windows.Media;

namespace NinjaCalc.Core {

    /// <summary>
    /// Encapsulates a validation result.
    /// </summary>
    public class CalcValidationLevel {

        public string Name {
            get;
            set;
        }

        public System.Windows.Media.Brush BorderBrush {
            get;
            set;
        }

        public System.Windows.Media.Brush BackgroundBrush {
            get;
            set;
        }

        public CalcValidationLevel(string name, System.Windows.Media.Brush borderBrush, System.Windows.Media.Brush backgroundBrush) {
            this.Name = name;
            this.BorderBrush = borderBrush;
            this.BackgroundBrush = backgroundBrush;
        }
    }

    /// <summary>
    /// The different ValidationLevels a validation function can return.
    /// </summary>
    public class CalcValidationLevels {
        public static readonly CalcValidationLevel Ok;
        public static readonly CalcValidationLevel Warning;
        public static readonly CalcValidationLevel Error;

        static CalcValidationLevels() {
            Ok = new CalcValidationLevel("ok", System.Windows.Media.Brushes.Green, (SolidColorBrush)new BrushConverter().ConvertFromString("#e5ffe5"));
            Warning = new CalcValidationLevel("warning", System.Windows.Media.Brushes.Orange, (SolidColorBrush)new BrushConverter().ConvertFromString("#fff5e5"));
            Error = new CalcValidationLevel("error", System.Windows.Media.Brushes.Red, (SolidColorBrush)new BrushConverter().ConvertFromString("#ffe5e5"));
        }
    }

    /// <summary>
    /// Designed to represent a single result from performing a validation on a calculator variable.
    /// </summary>
    public class CalcValidationResult {
        public CalcValidationLevel CalcValidationLevel {
            get;
            set;
        }

        public string Message {
            get;
            set;
        }

        public CalcValidationResult(CalcValidationLevel calcValidationLevel, string message) {
            this.CalcValidationLevel = calcValidationLevel;
            this.Message = message;
        }
    }

    /// <summary>
    /// Used to encapsulate a single validator for a calculator variable. Designed to be added to the CalcVar object.
    /// </summary>
    public class Validator {

        /// <summary>
        /// Gets or sets the validation function which performs the validation and returns a ValidationResult_t.
        /// </summary>
        public Func<double, CalcValidationLevel> ValidationFunction {
            get;
            set;
        }

        /// <summary>
        /// Gets or sets the message to be displayed to the user (in a tooltip), when the validator
        /// fails with either a warning or an error.
        /// </summary>
        public string Message {
            get;
            set;
        }

        /// <summary>
        /// Constructor. 
        /// </summary>
        /// <param name="validationFunction">The function which will validate the calculator variable.</param>
        public Validator(Func<double, CalcValidationLevel> validationFunction, string message) {
            // Save function internally
            this.ValidationFunction = validationFunction;

            // Save the message internally
            this.Message = message;
        }

        /// <summary>
        /// Factory function. Returns a validator which will give the provided "desiredValidationResult" if 
        /// the calculator variable is not a valid number (NaN does not count). If the value is a valid number,
        /// it will return "CalcValidationResults.Ok".
        /// </summary>
        /// <returns>A validator which will give an error if the calculator variable is not a valid number.</returns>
        public static Validator IsNumber(CalcValidationLevel desiredValidationResult) {
            return new Validator(
                (value) => {
                    //return ValidationResult_t.Error;
                    if (Double.IsNaN(value)) {
                        return desiredValidationResult;
                    }
                    else {
                        return CalcValidationLevels.Ok;
                    }
                },
                "Value must be a real number.");
        }

        /// <summary>
        /// Factory function. Returns a validator which will give the provided validation result if the calculator variable is 
        /// not greater than 0. If the number is greater than 0, it will return "CalcValidationResults.Ok".
        /// </summary>
        /// <returns>A validator which will give an error if the calculator variable is not a valid number.</returns>
        public static Validator IsGreaterThanZero(CalcValidationLevel desiredValidationResult) {
            return new Validator(
                (value) => {
                    //return ValidationResult_t.Error;
                    if (value <= 0) {
                        return desiredValidationResult;
                    }
                    else {
                        return CalcValidationLevels.Ok;
                    }
                },
                "Value must be positive and not equal to 0.");
        }
    }
}

File: src/Core/Validator.java
package Core;


import Core.CalcVar.CalcVarBase;

import java.util.ArrayList;

/**
 * Used to encapsulate a single validator for a calculator variable. Designed to be added to the CalcVarNumerical object
 * via addValidator().
 *
 * @author          gbmhunter <gbmhunter@gmail.com> (www.mbedded.ninja)
 * @since           2015-11-02
 * @last-modified   2016-04-13
 */
public class Validator {

    ArrayList<CalcVarBase> dependencies;

    /**
     * Gets or sets the validation function which performs the validation and returns a ValidationResult_t.
     */
    public IValidationFunctionWithValue ValidationFunction;

    /**
     * Gets or sets the message to be displayed to the user (in a tooltip), when the validator
     * fails with either a warning or an error.
     */
    public String Message;

    //===============================================================================================//
    //========================================== CONSTRUCTORS =======================================//
    //===============================================================================================//

    /**
     *  Constructor which accepts a validation function with the associated calculator variables raw value
     * as an input. Useful for creating pre-defined, simple validators.
     *
     * @param validationFunction    The function which will validate the calculator variable.
     * @param message
     */
    public Validator(IValidationFunctionWithValue validationFunction, String message) {

        this.dependencies = new ArrayList<>();

        // Save function internally
        this.ValidationFunction = validationFunction;

        // Save the message internally
        this.Message = message;
    }

    /**
     * Constructor which accepts a validation function with no input arguments. Useful for creating custom validators.
     *
     * @param validationFunction    The function which will validate the calculator variable.
     * @param message
     */
    public Validator(IValidationFunction validationFunction, String message) {

        this.dependencies = new ArrayList<>();

        // Save function internally
        this.ValidationFunction = (value) -> {
            CalcValidationLevel validationLevel = validationFunction.execute();
            return validationLevel;
        };

        // Save the message internally
        this.Message = message;
    }

    public Validator(ArrayList<CalcVarBase> dependencies, IValidationFunction validationFunction, String message) {

        // Save list of dependencies
        this.dependencies = dependencies;

        // Save function internally
        this.ValidationFunction = (value) -> {
            CalcValidationLevel validationLevel = validationFunction.execute();
            return validationLevel;
        };

        // Save the message internally
        this.Message = message;
    }

    //===============================================================================================//
    //========================================= FACTORY METHODS =====================================//
    //===============================================================================================//

    /**
     * Factory method. Returns a validator which will give the provided "desiredValidationResult" if
     * the calculator variable is not a valid number (NaN does not count). If the value is a valid number,
     * it will return "CalcValidationResults.Ok".
     *
     * @param   desiredValidationResult     The desired validation result if the calculator variable is not a number.
     * @return  A validator which will give the desired validation result if the calculator variable is not a valid number.
     */
    public static Validator IsNumber(CalcValidationLevel desiredValidationResult) {
        return new Validator(
            (value) -> {
                //return ValidationResult_t.Error;
                if (Double.isNaN(value)) {
                    return desiredValidationResult;
                }
                else {
                    return CalcValidationLevels.Ok;
                }
            },
            "Value must be a real number.");
    }

    /**
     * Factory function. Returns a validator which will give the provided validation result if the calculator variable is
     * not greater than 0. If the number is greater than 0, it will return "CalcValidationResults.Ok".
     *
     * @param desiredValidationResult   The desired validation result if the calculator variable value is not greater than 0.
     * @return  A validator which will give the desired validation result if the calculator variable is not greater than 0.
     */
    public static Validator IsGreaterThanZero(CalcValidationLevel desiredValidationResult) {
        return new Validator(
            (value) -> {
                //return ValidationResult_t.Error;
                if (value <= 0) {
                    return desiredValidationResult;
                }
                else {
                    return CalcValidationLevels.Ok;
                }
            },
            "Value must be positive and not equal to 0.");
    }
}
